internal/graph/graph.go

```diff
@@ -0,0 +1,112 @@
+/*
+Copyright The ORAS Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+// Package graph traverses graphs.
+package graph
+
+import (
+	"context"
+	"errors"
+
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+	"golang.org/x/sync/errgroup"
+	"golang.org/x/sync/semaphore"
+)
+
+// Dispatch traverses a graph concurrently. To maximize the concurrency, the
+// resulted search is neither depth-first nor breadth-first. For a rooted DAG,
+// the root node is always traversed first and then its down edges.
+// On visiting a node,
+// - `preHandler` is called before traversing the down edges.
+// - `postHandler` is called after traversing the down edges.
+// An optional concurrency limiter can be passed in to control the concurrency
+// level.
+// A handler may return `ErrSkipDesc` to signal not traversing descendants.
+// If any handler returns an error, the entire dispatch is cancelled.
+// This function is based on github.com/containerd/containerd/images.Dispatch.
+// Note: Handlers with `github.com/containerd/containerd/images.ErrSkipDesc`
+// cannot be used in this function.
+// WARNING:
+// - This function does not detect circles. It is possible running into an
```

Comment:
We should guard this as over time ORAS will be called in runtimes and not just CLIs. I would say we prioritize this or ensure you have an issue and let's get this fixed. We also use this in Ratify and so it would blow up the verifier if we see a graph that is messed up.

Reply:
This function is guard as an [internal](https://go.dev/doc/go1.4#internalpackages) package. The library consumers are not able to import this `internal/graph` pacakge. In other words, only ORAS can invoke this function, and Ratify cannot. As long as we have proper guarded in the exported methods of ORAS, we are fine.

In this PR, `Dispatch()` is invoked by `CopyGraph()` and `CopyGraph()` avoids circles.

Reply:
Ok sounds good. If we don’t have any more feedback would LGTM this change and keep making progress. 
I’ll ack by EOD PST.